File: C#/RankingApp/Ranking.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace RankingApp
{
    public class Ranking<T>
    {
        private List<(T item, int rank)> elements = new List<(T item, int rank)>();

        public void Add(T item, int rank)
        {
            elements.Add((item, rank));
            elements = elements.OrderBy(e => e.rank).ToList();
        }

        public List<T> GetTop(int count)
        {
            return elements.Take(count).Select(e => e.item).ToList();
        }
    }
}

File: C++/RankingApp/ranking.cpp
#include "ranking.h"

template <typename T>
void Ranking<T>::Add(T item, int rank) {
    items.push_back(std::make_pair(rank, item));
    std::sort(items.begin(), items.end(), [](const std::pair<int, T>& a, const std::pair<int, T>& b) {
        return a.first < b.first;
    });
}

template <typename T>
std::vector<T> Ranking<T>::GetTop(int count) const {
    std::vector<T> top_items;
    for (int i = 0; i < count && i < items.size(); ++i) {
        top_items.push_back(items[i].second);
    }
    return top_items;
}

template class Ranking<int>;
template class Ranking<std::string>;